src/engine/ult.c

```diff
@@ -374,30 +374,27 @@ static inline uint32_t
 sched_ult2xs_multisocket(int xs_type, int tgt_id)
 {
 	static __thread uint32_t offload;
-	uint32_t                 socket;
+	uint32_t                 socket = tgt_id / dss_tgt_per_numa_nr;
 	uint32_t                 base;
 	uint32_t                 target;
 
 	if (dss_tgt_offload_xs_nr == 0) {
 		if (xs_type == DSS_XS_IOFW && dss_forward_neighbor) {
 			/* Keep the old forwarding behavior, but NUMA aware */
-			socket = tgt_id / dss_numa_nr;
 			target = (socket * dss_tgt_per_numa_nr) +
 				 (tgt_id + offload) % dss_tgt_per_numa_nr;
-			offload = target + 17; /* Seed next selection */
-			target  = DSS_MAIN_XS_ID(target);
+			target = DSS_MAIN_XS_ID(target);
 			goto check;
 		}
 		return DSS_XS_SELF;
 	}
 
-	socket  = tgt_id / dss_numa_nr;
-	base    = dss_sys_xs_nr + dss_tgt_nr + (socket * dss_offload_per_numa_nr);
-	target  = base + ((offload + tgt_id) % dss_offload_per_numa_nr);
-	offload = target + 17; /* Seed next selection */
+	base   = dss_sys_xs_nr + dss_tgt_nr + (socket * dss_offload_per_numa_nr);
+	target = base + ((offload + tgt_id) % dss_offload_per_numa_nr);
```

Comment:
Do we need to keep the "target" and "tgt_id" within the same NUMA node or not? If yes, how? It seems that "base" may exceeds the boundary?

Reply:
Target in this calculation is a helper thread and will be on the same numa node according to the expected layout shown here: 

![image](https://github.com/daos-stack/daos/assets/10464486/2840c466-6232-495d-8895-97d63f7a829c)

Reply:
To be more precise, the helper threads are split among numa nodes but base is the xs id of the first helper thread.

Reply:
Thank for the detailed explanation.